### ISSUE TITLE
perf: split L6 data into separate obsolete, live files

### ISSUE THREAD

**Issue by jbowens** [cockroachdb/pebble#847]
In discussing compactions to remove obsolete keys from L6, Peter [had an idea](https://github.com/cockroachdb/pebble/pull/839#pullrequestreview-457499150):
> I wonder if we could arrange for any key written to a bottommost level that is pinned by a snapshot to be written to a separate sstable. That is, L6 would be separated into L6-live and L6-obsolete where L6-obsolete contains records that are pinned by a snapshot. With this setup, we simply have to wait until snapshots are released and then can perform a compaction which deletes the L6-obsolete sstable.

L6 files may contain obsolete records that must be preserved because they're pinned by an open snapshot. When the pinning snapshot eventually closes, we want to reclaim the disk space occupied by these obsolete records (#838), but L6 tables are large and expensive to compact. If the obsolete keys were segmented into a separate file, the obsolete file could cheaply be dropped by a simple manifest edit once the pinning snapshots were released. This would prevent unnecessary read and write IO.

The L6 obsolete files form an additional level beneath L6. Because records in the L6-obsolete level are _known_ to be obsolete (vs ordinary levels where records _might_ be obsolete), reads may skip any L6-obsolete files containing only sequence numbers strictly less than the iterator sequence number. Read amplification for reads at recent sequence numbers is unaffected, and range iterators at these recent sequence numbers avoid needing to skip over obsolete keys. Read amplification for reads at old sequence numbers is increased by 1.

We would need to experiment to get a sense of how much obsolete data is written to L6 sstables in practice to understand if such a large undertaking is worthwhile.

Jira issue: PEBBLE-215

**Comment by petermattis** [comment]
Cc @sumeerbhola in case you have thoughts about this. There is similarity to the idea of segregating historic MVCC versions, though I think this idea is simpler because it applies to the internal Pebble versions for a key. We could also consider generalizing this beyond L6. We only ever get multiple versions of a key in an sstable due to snapshots. We could consider always writing the obsolete versions to a "sidecar" sstable.

**Comment by petermattis** [comment]
PS Adding this functionality in a RocksDB-compatible way feels challenging.

**Comment by jbowens** [comment]
A good first step here might be to start calculating a new "obsolete entries" table property during compaction. The table property would help L6->L6 compaction picking (#838) determine whether it's worthwhile to compact a L6 file, even if the division of obsolete data into separate files doesn't prove worth pursuing.

**Comment by petermattis** [comment]
This idea is still valid. If snapshots are pinning data in L6 sstables this would give a faster path to deleting that data rather than having to perform a compaction involving the L6 sstable.

**Comment by sumeerbhola** [comment]
With EventuallyFileOnlySnapshot, such snapshot garbage should be minimal (though I don't know if we have looked at the before and after stats). Also, recently @jbowens removed all uses of the old snapshot from CockroachDB https://github.com/cockroachdb/cockroach/pull/142493

**Comment by petermattis** [comment]
Why do EFOS reduce obsolete records from snapshots? I thought EFOS prevented pinning memtables, but compactions still have to maintain old versions of records, right? Or do EFOS actually keep the old sstables around? (I'm unfamiliar with the implementation).

**Comment by jbowens** [comment]
Yeah, once the relevant memtables are flushed, the Version and its sstables get pinned. So the only obsolete records are those preserved during the flush.
